src/ir/child-typer.h

```diff
@@ -0,0 +1,1074 @@
+/*
+ * Copyright 2024 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef wasm_ir_child_typer_h
+#define wasm_ir_child_typer_h
+
+#include "wasm-traversal.h"
+#include "wasm.h"
+
+namespace wasm {
+
+// For each child, call `noteSubtype` with a pointer to the child and the most
+// specific type that the child must have.
+template<typename Subtype> struct ChildTyper : OverriddenVisitor<Subtype> {
+  Module& wasm;
+  Function* func;
+
+  ChildTyper(Module& wasm, Function* func) : wasm(wasm), func(func) {}
+
+  Subtype& self() { return *static_cast<Subtype*>(this); }
+
+  void note(Expression** childp, Type type) {
+    self().noteSubtype(childp, type);
+  }
+
+  void notePointer(Expression** ptrp, Name mem) {
+    note(ptrp, wasm.getMemory(mem)->indexType);
+  }
+
+  void noteAny(Expression** childp) { self().noteAnyType(childp); }
+
+  void noteAnyReference(Expression** childp) {
+    self().noteAnyReferenceType(childp);
+  }
+
+  void noteAnyTuple(Expression** childp, size_t arity) {
+    self().noteAnyTupleType(childp, arity);
+  }
+
+  Type getLabelType(Name label) { return self().getLabelType(label); }
+
+  void visitNop(Nop* curr) {}
+
+  void visitBlock(Block* curr) {
+    size_t n = curr->list.size();
+    for (size_t i = 0; i < n - 1; ++i) {
```

Comment:
When `n` is zero this is `i = 0; i < 0xffffff..; i++` which seems bad?

Early-exiting on top seems safer, and then can remove the condition on 61.